rust/noosphere-core/src/data/body_chunk.rs

```diff
@@ -70,4 +76,342 @@ impl BodyChunkIpld {
+        }
+
+        let mut store = self.store.to_owned();
+        let mut next_chunk_cid = None;
+        let chunk_stream = self.data.into_stream_rev().await;
+        tokio::pin!(chunk_stream);
+
+        while let Some(chunk) = chunk_stream.try_next().await? {
+            next_chunk_cid = Some(
+                store
+                    .save::<DagCborCodec, _>(&BodyChunkIpld {
+                        bytes: chunk,
+                        next: next_chunk_cid,
+                    })
+                    .await?,
+            );
+        }
+
+        match next_chunk_cid {
+            Some(cid) => Ok(cid),
+            None => Err(anyhow!("Could not encode empty buffer.")),
+        }
+    }
+}
+
+/// Container for short-term storage of sequenced blobs
+/// for consumption as a forward or reverse stream.
+/// [BlobVec] data is backed by memory, dynamically converting
+/// to [Scratch] provider storage if memory threshold is reached.
+/// Blobs can then be drained via `BlobVec::into_stream`.
+struct BlobVec<'a, S: Scratch> {
+    chunks: BlobVecData<S::ScratchStore>,
+    memory_limit: u64,
+    provider: &'a S,
+}
+
+impl<'a, S: Scratch> BlobVec<'a, S> {
+    /// Creates a new [BlobVec] that stores data in memory
+    /// until `memory_limit` is reached.
+    pub fn new(provider: &'a S, memory_limit: u64) -> Self {
+        BlobVec {
+            chunks: BlobVecData::new_with_memory(),
+            memory_limit,
+            provider,
+        }
+    }
+
+    /// Returns the number of items pushed.
+    pub fn len(&self) -> &usize {
+        self.chunks.chunk_count()
+    }
+
+    /// Push a new item.
+    pub async fn push(&mut self, chunk: Vec<u8>) -> Result<()> {
+        self.chunks.push(chunk).await?;
+
+        if matches!(self.chunks, BlobVecData::Memory { .. })
+            && self.chunks.byte_length() > &self.memory_limit
+        {
+            self.upgrade_storage().await?;
+        }
+
+        Ok(())
+    }
+
+    /// Drain all items into a stream.
+    #[allow(unused)]
+    pub async fn into_stream(self) -> impl Stream<Item = Result<Vec<u8>, anyhow::Error>> + Unpin {
+        self.chunks.into_stream(false).await
+    }
+
+    /// Drain all items into a stream in reverse.
+    pub async fn into_stream_rev(
+        self,
+    ) -> impl Stream<Item = Result<Vec<u8>, anyhow::Error>> + Unpin {
+        self.chunks.into_stream(true).await
+    }
+
+    async fn upgrade_storage(&mut self) -> Result<()> {
+        let store = self.provider.get_scratch_store().await?;
+        let mut storage = BlobVecData::new_with_db(store).await?;
+        std::mem::swap(&mut self.chunks, &mut storage);
+        let previous_stream = storage.into_stream(false).await;
+        tokio::pin!(previous_stream);
+        while let Some(chunk) = previous_stream.try_next().await? {
+            self.chunks.push(chunk).await?;
+        }
+        Ok(())
+    }
+}
+
+enum BlobVecData<S: KeyValueStore> {
```

Comment:
I would say that the private implementation within this module deserves at least some top-level documentation for legibility's sake.